sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -30,13 +30,16 @@ case class ScalaUDF(
     function: AnyRef,
     dataType: DataType,
     children: Seq[Expression],
-    inputTypes: Seq[DataType] = Nil)
+    inputTypes: Seq[DataType] = Nil,
+    isDeterministic: Boolean = true)
   extends Expression with ImplicitCastInputTypes with CodegenFallback {
 
   override def nullable: Boolean = true
 
   override def toString: String = s"UDF(${children.mkString(",")})"
 
+  override def deterministic: Boolean = isDeterministic && children.forall(_.deterministic)
```

Comment:
We probably need to override the `foldable` as well. Something like:
```scala
override def foldable: Boolean = deterministic && children.forall(_.foldable)
```
Or change it in `Expression` directly?